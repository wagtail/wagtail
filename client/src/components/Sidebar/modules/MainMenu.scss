// stylelint-disable declaration-no-important
.sidebar-main-menu {
  overflow: auto;
  overflow-x: hidden;

  &__list {
    margin: 0;
    padding: 0;
    list-style-type: none;
  }

  *:focus {
    @include show-focus-outline-inside;
  }

  .icon--submenu-header {
    display: block;
    width: 4rem;
    height: 4rem;
    margin: 0 auto 0.8em;
    opacity: 0.15;
  }

  > ul > li > a {
    // Need !important to override body.ready class
    transition: padding $menu-transition-duration ease !important;

    .menuitem-label {
      transition: opacity $menu-transition-duration ease;
    }
  }
}

.sidebar-footer {
  @apply w-bg-surface-menus w-mt-auto;
  transition: width $menu-transition-duration ease !important; // Override body.ready

  > ul,
  ul > li {
    margin: 0;
    padding: 0;
    list-style-type: none;
  }

  ul > li {
    position: relative;
  }

  > ul {
    @include transition(max-height $menu-transition-duration ease);
    visibility: hidden;

    max-height: 0;

    a {
      border-inline-start: 3px solid transparent;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;

      &:before {
        font-size: 1rem;
        margin-inline-end: 0.5em;
        vertical-align: -10%;
      }
    }
  }

  &__account {
    @include show-focus-outline-inside();

    &-toggle {
      @apply w-pl-2 w-inline-flex w-justify-between w-w-full w-translate-x-0 w-transition w-duration-150;

      min-width: 0;

      .icon {
        flex-shrink: 0;
      }
    }

    &-label {
      @apply w-text-text-label-menus-default w-text-left w-overflow-hidden w-whitespace-nowrap w-text-ellipsis;
    }
  }

  @at-root .sidebar--slim #{&} {
    &__account {
      @apply w-pb-3;
    }

    &__account-toggle {
      @apply w-hidden;
    }
  }

  &--open {
    > ul {
      $footer-submenu-height: 85px;
      max-height: $footer-submenu-height;
      visibility: visible;
    }
  }
}
