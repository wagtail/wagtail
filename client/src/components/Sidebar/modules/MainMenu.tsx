import * as React from 'react';
import Icon from '../../Icon/Icon';

import { LinkMenuItemDefinition } from '../menu/LinkMenuItem';
import { MenuItemDefinition } from '../menu/MenuItem';
import { SubMenuItemDefinition } from '../menu/SubMenuItem';
import { ModuleDefinition, Strings } from '../Sidebar';

export function renderMenu(
  path: string,
  items: MenuItemDefinition[],
  slim: boolean,
  state: MenuState,
  dispatch: (action: MenuAction) => void,
  navigate: (url: string) => Promise<void>
) {
  return (
    <>
      {items.map(item => item.render({
        path: `${path}.${item.name}`,
        slim,
        state,
        dispatch,
        navigate,
      }))}
    </>
  );
}

interface SetActivePath {
  type: 'set-active-path',
  path: string,
}

interface SetNavigationPath {
  type: 'set-navigation-path',
  path: string,
}

export type MenuAction = SetActivePath | SetNavigationPath;

export interface MenuState {
  navigationPath: string;
  activePath: string;
}

function menuReducer(state: MenuState, action: MenuAction) {
  const newState = Object.assign({}, state);

  if (action.type === 'set-active-path') {
    newState.activePath = action.path;
  } else if (action.type === 'set-navigation-path') {
    newState.navigationPath = action.path;
  }

  return newState;
}

interface MenuProps {
  menuItems: MenuItemDefinition[];
  accountMenuItems: MenuItemDefinition[];
  user: MainMenuModuleDefinition['user'];
  slim: boolean;
  expandingOrCollapsing: boolean;
  currentPath: string;
  strings: Strings;
  navigate(url: string): Promise<void>;
}

export const Menu: React.FunctionComponent<MenuProps> = (
  { menuItems, accountMenuItems, user, expandingOrCollapsing, slim, currentPath, strings, navigate }) => {
  // navigationPath and activePath are two dot-delimited path's referencing a menu item
  // They are created by concatenating the name fields of all the menu/sub-menu items leading to the relevant one.
  // For example, the "Users" item in the "Settings" sub-menu would have the path 'settings.users'
  // - navigationPath references the current sub-menu that the user currently has open
  // - activePath references the menu item for the the page the user is currently on
  const [state, dispatch] = React.useReducer(menuReducer, {
    navigationPath: '',
    activePath: '',
  });
  const accountSettingsOpen = state.navigationPath.startsWith('.account');

  // Whenever currentPath or menu changes, work out new activePath
  React.useEffect(() => {
    const urlPathsToNavigationPaths: [string, string][] = [];
    const walkMenu = (path: string, walkingMenuItems: MenuItemDefinition[]) => {
      walkingMenuItems.forEach((item) => {
        const newPath = `${path}.${item.name}`;

        if (item instanceof LinkMenuItemDefinition) {
          urlPathsToNavigationPaths.push([item.url, newPath]);
        } else if (item instanceof SubMenuItemDefinition) {
          walkMenu(newPath, item.menuItems);
        }
      });
    };

    walkMenu('', menuItems);

    let bestMatch: [string, string] | null = null;
    urlPathsToNavigationPaths.forEach(([urlPath, navPath]) => {
      if (currentPath.startsWith(urlPath)) {
        if (bestMatch == null || urlPath.length > bestMatch[0].length) {
          bestMatch = [urlPath, navPath];
        }
      }
    });

    const newActivePath = bestMatch ? bestMatch[1] : '';
    if (newActivePath !== state.activePath) {
      dispatch({
        type: 'set-active-path',
        path: newActivePath,
      });
    }
  }, [currentPath, menuItems]);

  React.useEffect(() => {
    // Close submenus when user presses escape
    const onKeydown = (e: KeyboardEvent) => {
      // IE11 uses "Esc" instead of "Escape"
      if (e.key === 'Escape' || e.key === 'Esc') {
        dispatch({
          type: 'set-navigation-path',
          path: ''
        });
      }
    };

    document.addEventListener('keydown', onKeydown);

    return () => {
      document.removeEventListener('keydown', onKeydown);
    };
  }, []);

  // Whenever the parent Sidebar component collapses or expands, close any open menus
  React.useEffect(() => {
    if (expandingOrCollapsing) {
      dispatch({
        type: 'set-navigation-path',
        path: ''
      });
    }
  }, [expandingOrCollapsing]);

  const onClickAccountSettings = (e: React.MouseEvent) => {
    e.preventDefault();

    if (accountSettingsOpen) {
      dispatch({
        type: 'set-navigation-path',
        path: '',
      });
    } else {
      dispatch({
        type: 'set-navigation-path',
        path: '.account',
      });
    }
  };

  const className = (
    'sidebar-main-menu'
    + (accountSettingsOpen ? ' sidebar-main-menu--open-footer' : '')
  );

  return (
    <>
      <nav className={className}>
        <ul className="sidebar-main-menu__list">
          {renderMenu('', menuItems, slim, state, dispatch, navigate)}
        </ul>
      </nav>
      <div className={'sidebar-footer' + (accountSettingsOpen ? ' sidebar-footer--open' : '')}>
        <button
          className="sidebar-footer__account"
          title={strings.EDIT_YOUR_ACCOUNT}
          aria-label={strings.EDIT_YOUR_ACCOUNT}
          onClick={onClickAccountSettings}
          aria-haspopup="true"
          aria-expanded={accountSettingsOpen ? 'true' : 'false'}
        >
          <div className="avatar square avatar-on-dark">
            <img src={user.avatarUrl} alt="" />
          </div>
          <div className="sidebar-footer__account-toggle">
            <div className="sidebar-footer__account-label">{user.name}</div>
            <Icon
              className="sidebar-footer__account-icon"
              name={(accountSettingsOpen ? 'arrow-down' : 'arrow-up')}
            />
          </div>
        </button>

        <ul>
          {renderMenu('', accountMenuItems, slim, state, dispatch, navigate)}
        </ul>
      </div>
    </>
  );
};

export class MainMenuModuleDefinition implements ModuleDefinition {
  menuItems: MenuItemDefinition[];
  accountMenuItems: MenuItemDefinition[];
  user: {
      name: string;
      avatarUrl: string;
  };

  constructor(
    menuItems: MenuItemDefinition[],
    accountMenuItems: MenuItemDefinition[],
    user: MainMenuModuleDefinition['user']
  ) {
    this.menuItems = menuItems;
    this.accountMenuItems = accountMenuItems;
    this.user = user;
  }

  render({ slim, expandingOrCollapsing, key, currentPath, strings, navigate }) {
    return (
      <Menu
        menuItems={this.menuItems}
        accountMenuItems={this.accountMenuItems}
        user={this.user}
        slim={slim}
        expandingOrCollapsing={expandingOrCollapsing}
        key={key}
        currentPath={currentPath}
        strings={strings}
        navigate={navigate}
      />
    );
  }
}
