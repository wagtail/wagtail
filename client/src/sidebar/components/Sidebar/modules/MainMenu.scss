// stylelint-disable declaration-no-important
.sidebar-main-menu {
  overflow: auto;
  overflow-x: hidden;
  // So the last items in the menu will be seen when the menu is vertically scrollable
  margin-bottom: 52px;

  @include transition(margin-bottom $menu-transition-duration ease);

  &--open-footer {
    margin-bottom: 127px;
  }

  &__list {
    margin: 0;
    padding: 0;
    list-style-type: none;
  }

  *:focus {
    @include show-focus-outline-inside;
  }

  .icon--submenu-header {
    display: block;
    width: 4rem;
    height: 4rem;
    margin: 0 auto 0.8em;
    opacity: 0.15;
  }

  > ul > li > a {
    // Need !important to override body.ready class
    transition: padding $menu-transition-duration ease !important;

    .menuitem-label {
      transition: opacity $menu-transition-duration ease;
    }
  }
}

.sidebar-footer {
  @apply w-bg-primary w-bottom-0 w-fixed;
  transition: width $menu-transition-duration ease !important; // Override body.ready
  width: $menu-width;

  > ul,
  ul > li {
    margin: 0;
    padding: 0;
    list-style-type: none;
  }

  ul > li {
    position: relative;

    @include transition(border-color $menu-transition-duration ease);
  }

  > ul {
    @include transition(max-height $menu-transition-duration ease);
    visibility: hidden;

    max-height: 0;

    a {
      border-inline-start: 3px solid transparent;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;

      &:before {
        font-size: 1rem;
        margin-inline-end: 0.5em;
        vertical-align: -10%;
      }
    }
  }

  &__account {
    @include show-focus-outline-inside();

    &-toggle {
      @apply w-pl-2 w-inline-flex w-justify-between w-w-full w-translate-x-0 w-transition w-duration-150;
    }

    &-label {
      @apply w-text-white w-text-left w-overflow-hidden w-whitespace-nowrap w-text-ellipsis;
    }
  }

  @at-root .sidebar--slim #{&} {
    width: $menu-width-slim;

    &__account {
      @apply w-px-0 w-pb-3 w-justify-center;
    }

    &__account-toggle {
      @apply w-hidden;
    }
  }

  &--open {
    width: $menu-width !important; // Override collapsed style

    > ul {
      $footer-submenu-height: 85px;
      max-height: $footer-submenu-height;
      visibility: visible;
    }
  }
}
