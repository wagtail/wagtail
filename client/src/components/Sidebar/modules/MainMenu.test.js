import React from 'react';
import { shallow } from 'enzyme';
import { Menu } from './MainMenu';

describe('Menu', () => {
  const user = { avatarUrl: 'https://gravatar/profile' };

  it('should render with the minimum required props', () => {
    const wrapper = shallow(
      <Menu accountMenuItems={[]} menuItems={[]} user={user} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the sidebar footer (account) when clicked', () => {
    const wrapper = shallow(
      <Menu accountMenuItems={[]} menuItems={[]} user={user} />,
    );

    // default is closed
    expect(wrapper.find('.sidebar-footer__account').prop('aria-expanded')).toBe(
      'false',
    );
    expect(wrapper.find('.sidebar-footer--open')).toHaveLength(0);

    wrapper.find('.sidebar-footer__account').simulate('click');

    expect(wrapper.find('.sidebar-footer__account').prop('aria-expanded')).toBe(
      'true',
    );
    expect(wrapper.find('.sidebar-footer--open')).toHaveLength(1);
  });
});
