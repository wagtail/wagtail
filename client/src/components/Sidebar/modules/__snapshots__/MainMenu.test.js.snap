// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render with the minimum required props 1`] = `
<Fragment>
  <nav
    aria-label="Main menu"
    className="sidebar-main-menu w-scrollbar-thin"
  >
    <ul
      className="sidebar-main-menu__list"
    />
  </nav>
  <div
    className="sidebar-footer sidebar-footer--visible"
  >
    <ForwardRef(TippyWrapper)
      disabled={true}
      placement="right"
    >
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Edit your account"
        className="
            w-px-5
            sidebar-footer__account
            w-bg-surface-menus
            w-text-text-label-menus-default
            w-flex
            w-items-center
            w-relative
            w-w-full
            w-appearance-none
            w-border-0
            w-overflow-hidden
            w-py-3
            hover:w-bg-surface-menu-item-active
            focus:w-bg-surface-menu-item-active
            w-transition"
        onClick={[Function]}
        title="Edit your account"
        type="button"
      >
        <div
          className="avatar avatar-on-dark w-flex-shrink-0 !w-w-[28px] !w-h-[28px]"
        >
          <img
            alt=""
            decoding="async"
            loading="lazy"
            src="https://gravatar/profile"
          />
        </div>
        <div
          className="sidebar-footer__account-toggle"
        >
          <div
            className="sidebar-footer__account-label w-label-3"
          />
          <Icon
            className="w-w-4 w-h-4 w-text-text-label-menus-default"
            name="arrow-up"
          />
        </div>
      </button>
    </ForwardRef(TippyWrapper)>
    <ul />
  </div>
</Fragment>
`;
